tests/decomposition/test_decomposition_graph.py

```diff
@@ -341,3 +341,121 @@ def custom_controlled_decomp(wires):
+        assert len(graph._graph.nodes()) == 3
+        assert len(graph._graph.edges()) == 2
+
+        graph.solve()
+        with qml.queuing.AnnotatedQueue() as q:
+            graph.decomposition(op)(*op.parameters, wires=op.wires, **op.hyperparameters)
+
+        assert q.queue == [qml.RX(-0.5, wires=[0])]
+        assert graph.resource_estimate(op) == to_resources({qml.RX: 1})
+
+    def test_controlled_adjoint(self, _):
+        """Tests that the adjoint of a controlled operator is decomposed properly."""
+
+        op = qml.adjoint(qml.ops.Controlled(qml.RX(0.5, wires=[0]), control_wires=1))
+        op2 = qml.adjoint(qml.ctrl(qml.U1(0.5, wires=0), control=[1]))
+
+        graph = DecompositionGraph(
+            operations=[op, op2], target_gate_set={"ControlledPhaseShift", "CRX"}
+        )
+        # 5 operator nodes: Adjoint(C(RX)), Adjoint(C(U1)), CRX, C(U1), ControlledPhaseShift
+        # 3 decomposition nodes leading into Adjoint(C(RX)), Adjoint(C(U1)), and C(U1)
+        assert len(graph._graph.nodes()) == 8
+        # 3 edges from decompositions to ops and 3 edges from ops to decompositions
+        assert len(graph._graph.edges()) == 6
+
+        graph.solve()
+        with qml.queuing.AnnotatedQueue() as q:
+            graph.decomposition(op)(*op.parameters, wires=op.wires, **op.hyperparameters)
+            graph.decomposition(op2)(*op2.parameters, wires=op2.wires, **op2.hyperparameters)
+
+        assert q.queue == [qml.CRX(-0.5, wires=[1, 0]), qml.ctrl(qml.U1(-0.5, wires=0), control=1)]
+
+        op3 = qml.ctrl(qml.U1(-0.5, wires=0), control=1)
+        with qml.queuing.AnnotatedQueue() as q:
+            graph.decomposition(op3)(*op3.parameters, wires=op3.wires, **op3.hyperparameters)
+
+        assert q.queue == [qml.ControlledPhaseShift(-0.5, wires=[1, 0])]
+        assert graph.resource_estimate(op2) == to_resources({qml.ControlledPhaseShift: 1})
+
+    def test_adjoint_general(self, _):
+        """Tests decomposition of a generalized adjoint operation."""
+
+        class CustomOp(qml.operation.Operation):  # pylint: disable=too-few-public-methods
+            """A custom operation."""
+
+            resource_keys = set()
+
+            @property
+            def resource_params(self):
+                return {}
+
+        @qml.register_resources({qml.H: 1, qml.CNOT: 2, qml.RX: 1, qml.T: 1})
+        def custom_decomp(phi, wires):
+            qml.H(wires[0])
+            qml.CNOT(wires=wires[:2])
+            qml.RX(phi, wires=wires[1])
+            qml.CNOT(wires=wires[1:3])
+            qml.T(wires[2])
+
+        op = qml.adjoint(CustomOp(0.5, wires=[0, 1, 2]))
+        graph = DecompositionGraph(
+            operations=[op],
+            target_gate_set={"Hadamard", "CNOT", "RX", "PhaseShift"},
+            alt_decomps={CustomOp: [custom_decomp]},
+        )
+        # 10 operator nodes: A(CustomOp), A(H), A(CNOT), A(RX), A(T), H, CNOT, RX, A(PhaseShift), PhaseShift
```

Comment:
The decomposition indicated by this series of ops makes me wonder:
It feels  a little inconsistent to not have A(T) and T be the same resource count (we did not add T to the list of ops that have this property), but then once we move to `PhaseShift`, this differentiation is lost.
Now, I think this is fine and one can argue that introducing the rule `T -> PhaseShift` is the problematic ("test user-defined") bit here. But I wanted to bring it up, wondering whether this structure could lead to unwanted simplifications in more complex scenarios.

Reply:
In general `A(Op)` is not the same as `Op`. `A(T)` being equivalent to `T` is in essence a decomposition. I try not to imbed any decomposition logic in the construction of the `CompressedResourceOp`.

Reply:
Yeah okay, that would be my take as well! So here the effective "A(T)=T" resulting from "A(PhaseShift)=PhaseShift" is considered a user-introduced equality, and it is fine because a user would have to introduce it themselves conciously. Got it :)